compendium/DeclarativeServices/src/metadata/MetadataParserImpl.cpp

```diff
@@ -135,6 +138,81 @@ MetadataParserImplV1::CreateComponentMetadata(const AnyMap& metadata) const
   ObjectValidator(metadata, "name", /*isOptional=*/true)
     .AssignValueTo(compMetadata->name);
 
+  // component.configuration-policy (Optional)
+  compMetadata->configurationPolicy = CONFIG_POLICY_IGNORE;
+  unsigned short configPolicy = 0;
+  object =
+    ObjectValidator(metadata, "configuration-policy", /*isOptional=*/true);
+  if (object.KeyExists()) {
+    object.AssignValueTo(compMetadata->configurationPolicy);
+    configPolicy = 1;
+  }
+
+  // component.configuration-pid (Optional)
+  unsigned short configPid = 0;
+  object = ObjectValidator(metadata, "configuration-pid", /*isOptional=*/true);
+  if (object.KeyExists()) {
+    configPid = 1;
+    if (configPolicy == 1) {
+      const auto configPids =
+        object.GetValue<std::vector<cppmicroservices::Any>>();
+      std::transform(
+        std::begin(configPids),
+        std::end(configPids),
+        std::back_inserter(compMetadata->configurationPids),
+        [](const auto& configPid) {
+          return ObjectValidator(configPid).GetValue<std::string>();
+        });
+
+      //search for a configuration pid equal to $. If present replace with component name.
+      // Also search for duplicates pids. These are errors.
+      std::unordered_map<std::string, std::string> duplicatePids;
+      for (auto& pid : compMetadata->configurationPids) {
+        if (pid == "$") {
+          pid = compMetadata->name;
+        }
+
+        if (duplicatePids.find(pid) != duplicatePids.end()) {
+          std::string msg =
+            "configuration-pid error in the manifest. Duplicate pid detected. ";
+          msg.append(pid);
+          throw std::runtime_error(msg);
+        }
+        duplicatePids.emplace(pid, pid);
+      };
+    }
```

Comment:
Jeff/Patty: move this code into source code generator (fail earlier when possible)
Jeff: if OSGi spec doesn't say anything explicitly re: duplicate PID, keep this logic but do not throw (just log as a warning).
Jeff/Patty: if this runs only when bundle is installed (not when started), this code can be removed.

Reply:
Source code generator now checks for duplicate pids. The build will fail if duplicate pids are detected in the configuration-pid attribute. 
The code in DS that parses the metadata runs when a bundle is started. This parsing code still checks for duplicate pids (I didn't remove it when I moved  functionality to source code generator) and logs an error and throws an exception if duplicate pids are detected.